kinesis/kinesis.go

```diff
@@ -637,6 +647,81 @@ func gzipCompress(data []byte) ([]byte, error) {
 	return b.Bytes(), nil
 }
 
+// Compress Then Truncate
+// compresses data with CompressorFunction and iteratively truncates data
+// adding the truncation suffix if the CompressorFunction output exceeds maxOutLen.
+// The output is compressed and possibly truncated data whose length guaranteed to
+// be less than or equal to maxOutLen.
+func compressThenTruncate(compressorFunc CompressorFunc, data []byte, maxOutLen int, truncatedSuffix []byte, outputPlugin OutputPlugin) ([]byte, error) {
+	var compressedData []byte
+	var truncationBuffer []byte
+	var originalCompressedLen int
+	var compressedLen int
+	var err error
+
+	/* Iterative approach to truncation */
+	isTruncated := false
+	compressedLen = math.MaxInt64
+	truncatedInLen := len(data)
+	truncationBuffer = data
+	for (maxOutLen < compressedLen) {
+		compressedData, err = compressorFunc(truncationBuffer)
+		if err != nil {
+			return nil, err
+		}
+		compressedLen = len(compressedData)
+
+		/* Truncation needed */
+		if (maxOutLen < compressedLen) {
+			logrus.Debugf("[kinesis %d] iterative truncation round stream=%s\n",
+						 outputPlugin.PluginID, outputPlugin.stream)
+
+			/* Base case: input compressed empty string, output still too large */
+			if (truncatedInLen == 0) {
+				logrus.Errorf("[kinesis %d] truncation failed, compressed empty input too " +
+							 "large stream=%s\n", outputPlugin.PluginID, outputPlugin.stream)
+				return nil, errors.New("compressed empty to large");
+			}
```

Comment:
Is this actually possible? The compressed length is too high by the data input size is 0? Also why do you call it the base case? This isn't a recursive function.

Reply:
Oh I see, this happens when you truncate the log until nothing is left? But then, how would the compressed length still be too large?